vpn/tunnel.go

```diff
@@ -327,19 +363,109 @@ func convertWorkspaceUpdate(update tailnet.WorkspaceUpdate) *PeerUpdate {
 	for i, agent := range update.DeletedAgents {
 		fqdn := make([]string, 0, len(agent.Hosts))
 		for name := range agent.Hosts {
-			fqdn = append(fqdn, name.WithoutTrailingDot())
+			fqdn = append(fqdn, name.WithTrailingDot())
 		}
 		out.DeletedAgents[i] = &Agent{
+			Id:            tailnet.UUIDToByteSlice(agent.ID),
+			Name:          agent.Name,
+			WorkspaceId:   tailnet.UUIDToByteSlice(agent.WorkspaceID),
+			Fqdn:          fqdn,
+			IpAddrs:       hostsToIPStrings(agent.Hosts),
+			LastHandshake: nil,
+			Latency:       nil,
+		}
+	}
+	return out, nil
+}
+
+// Given a list of agents, populate their network info, and return them as proto agents.
+func (t *Tunnel) populateAgents(agents []*tailnet.Agent) ([]*Agent, error) {
+	if t.conn == nil {
+		return nil, xerrors.New("no active connection")
+	}
+
+	out := make([]*Agent, 0, len(agents))
+	var wg sync.WaitGroup
+	pingCtx, cancelFunc := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancelFunc()
+
+	for _, agent := range agents {
+		fqdn := make([]string, 0, len(agent.Hosts))
+		for name := range agent.Hosts {
+			fqdn = append(fqdn, name.WithTrailingDot())
+		}
+		protoAgent := &Agent{
 			Id:          tailnet.UUIDToByteSlice(agent.ID),
 			Name:        agent.Name,
 			WorkspaceId: tailnet.UUIDToByteSlice(agent.WorkspaceID),
 			Fqdn:        fqdn,
-			IpAddrs:     []string{tailnet.CoderServicePrefix.AddrFromUUID(agent.ID).String()},
-			// TODO: Populate
-			LastHandshake: nil,
+			IpAddrs:     hostsToIPStrings(agent.Hosts),
 		}
+		agentIP := tailnet.CoderServicePrefix.AddrFromUUID(agent.ID)
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			duration, _, _, err := t.conn.Ping(pingCtx, agentIP)
```

Comment:
Let's leave out latency for the moment. I know we want it in the Desktop UX, but this implementation is half-baked. Doing it here means we're synchronously waiting for up to 5 seconds for pings while we are trying to populate an update message. In particular, this gets called while handling the GetPeerUpdate RPC, which means it blocks the entire `requestLoop()`.

It might make more sense to let the Desktop app directly ping the agents over the TUN in a background thread. That would allow us to report a more "end to end" latency value anyway.